### PR TITLE
Port foreman_environment to Apypie

### PR DESCRIPTION
**Pull request by ephracis**
Port the environment module to the new Apypie boilerplate. Until we
get #291 merged this means that we cannot modify associated orgs and
locs which is why those tests are commented out for now.

This is part of #274

### CODE REVIEW

**Comment by evgeni** [comment]
291 is IN! :)

**Comment by ephracis** [comment]
Travis fails on `test_crud[domain]` even though I haven't touched it.

**Comment by mdellweg** [comment]
@ephracis Yes, i think i experienced the same. I think i do not like to _just_ add the `domain-0.yaml` fixture. I would commit the complete rerecorded playbook.

**Comment by ephracis** [comment]
> @ephracis Yes, i think i experienced the same. I think i do not like to _just_ add the `domain-0.yaml` fixture. I would commit the complete rerecorded playbook.

Unfortunately I never get the initial call to `/apidoc/v2.json` even when I delete the cache. So if I do a rerecord and push it, it will fail. No matter what I do.

But it seems to be working fine to just copy it from another fixture.

**Comment by mdellweg** [comment]
Did you rebase on top of #291 ? The use of a tmpdir there should do exactly that.

**Comment by ephracis** [comment]
Yes I did and I did a rerecord och both environment and domain but still no such call. Perhaps there's something wrong with my setup (since it used to work before, as long as I cleared the cache). Though when I run `make test` it will attempt to do the request and I will get a failure since the `make record_xxx` did not properly record the request.

Very strange. I'll have to look into it later when I have the time. Or just nuke the setup and build a new from scratch perhaps.

**Comment by mdellweg** [comment]
Thanks for looking into this. Since i also already started to try to rerecord those (they are failing on master too), i will give it a shot.

**Comment by evgeni** [comment]
Maybe I broke it, as master is red too :(

**Comment by mdellweg** [comment]
I can confirm, that the first fixture does not contain the apidoc file.

**Comment by mdellweg** [comment]
@ephracis Tests on master should be fixed. Can you rebase?

**Comment by mdellweg** [comment]
https://travis-ci.org/theforeman/foreman-ansible-modules/jobs/541674163#L439

That directory does not really have the random name i'd expect from a tmpdir. Can you make sure it is empty?

**Comment by evgeni** [comment]
It should be empty as it's freshly created by pytest. But yes the name is not really random.

**Comment by ephracis** [comment]
Still not getting any proper apypie requests in the `-0` fixture. Is there anything I need to do besides rebasing?

**Comment by mdellweg** [comment]
I think, i know... You need to do set the XDG_..._CACHE_SOMETHING environment variable. Look at the new activationkey playbook.

**Comment by mdellweg** [comment]
```diff
diff --git a/test/test_playbooks/environment.yml b/test/test_playbooks/environment.yml
index bec0328..f4d529e 100644
--- a/test/test_playbooks/environment.yml
+++ b/test/test_playbooks/environment.yml
@@ -14,6 +14,8 @@
 
 - hosts: tests
   gather_facts: false
+  environment:
+    XDG_CACHE_HOME: "{{ lookup('env', 'FAM_TEST_APYPIE_CACHE_DIR') }}"
   tasks:
   - name: 'Load server config'
     include_vars:
```

**Comment by evgeni** [comment]
That looks good, yeah. We probably should write a short porting guide.

**Comment by mdellweg** [comment]
I guess, we wanted to change that anyway with the cache prewarming.

**Comment by ephracis** [comment]
That seems to have done the trick. Now I ran both a rerecord and a test locally and it worked fine. Hope it works in Travis as well.

File: plugins/modules/foreman_environment.py [hunk 1]
```diff
@@ -120,22 +106,15 @@ def main():
 
     module.connect()
 
-    try:
-        entity = find_environment(module, name=entity_dict['name'], failsafe=True)
-    except Exception as e:
-        module.fail_json(msg='Failed to find entity: %s ' % e)
+    entity = module.find_resource_by_name('environments', name=entity_dict['name'], failsafe=True)
 
-    # Set Locations of partition table
```
Comment (mdellweg) [inline plugins/modules/foreman_environment.py]: :+1:

File: plugins/modules/foreman_environment.py [hunk 2]
```diff
@@ -78,36 +79,21 @@
...
-        Location,
-        Organization,
-    )
-except ImportError:
-    pass
-
-from ansible.module_utils.foreman_helper import ForemanEntityAnsibleModule
-
+from ansible.module_utils.foreman_helper import (
+    sanitize_entity_dict,
```
Comment (mdellweg) [inline plugins/modules/foreman_environment.py]: You are not using this anymore.

File: plugins/modules/foreman_environment.py [hunk 3]
```diff
@@ -120,22 +106,15 @@ def main():
...
-    except Exception as e:
-        module.fail_json(msg='Failed to find entity: %s ' % e)
+    entity = module.find_resource_by_name('environments', name=entity_dict['name'], failsafe=True)
 
-    # Set Locations of partition table
     if 'locations' in entity_dict:
-        entity_dict['locations'] = find_locations(module, entity_dict['locations'])
+        entity_dict['locations'] = module.find_resources('locations', entity_dict['locations'], thin=True)
 
-    # Set Organizations of partition table
```
Comment (mdellweg) [inline plugins/modules/foreman_environment.py]: Those have still hung around?

**Comment by mdellweg** [comment]
somehow, your api.json still slips through. The first request in `fixtures/*-0.yml` must be the one to `apidoc/v2.json`.

**Comment by ephracis** [comment]
> somehow, your api.json still slips through. The first request in `fixtures/*-0.yml` must be the one to `apidoc/v2.json`.

Yeah, I just saw that. :(

**Comment by ephracis** [comment]
My cache path is `/Users/chbr/.cache/apypie/https___foreman.example.com/v2/default.json` and the cache file is created during the fixtures (when creating the org). So when it gets to the first environment call the cache file is present and the request is never made.

Note that I am running on a Mac. Maybe that would be a reason why it's not working for me?

**Comment by ephracis** [comment]
or that I have an old version of apypie installed...

**Comment by mdellweg** [comment]
And the environment variable is trying to move the cache for the tests host into to pytest tmpdir. Maybe that's what does not work in your environment.
Can you debug this by adding a module.fail_json with the env-var?
If nothing else, maybe running it inside a linux vm works?

**Comment by ephracis** [comment]
My api.py has not mention of the XDG variable. So it's ignored. I compared to what's on GitHub and it's different.

I had to uninstall it to make `make test-setup` install the correct version. Now I get a proper fixture file with the apidoc request.

**Comment by mdellweg** [comment]
`pip install --upgrade apypie` should have done the trick, too.

**Comment by evgeni** [comment]
Not sure upgrade works with git urls if the version doesn't change.

**Comment by mdellweg** [comment]
2 Acks. Let's get this one off your back. ;)

**Comment by mdellweg** [comment]
@ephracis Thank you for so much patience.

**Pull request merged**